src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,517 @@
+            // Due to WCF DS V2 compatibility we must not write the start boundary in this case
+            // otherwise WCF DS V2 won't be able to read it (it fails on the start-end boundary empty changeset).
+
+            // write the end boundary for the change set
+            ODataMultipartMixedBatchWriterUtils.WriteEndBoundary(this.rawOutputContext.TextWriter, currentChangeSetBoundary, !this.changesetStartBoundaryWritten);
+        }
+
+        /// <summary>
+        /// Creates an <see cref="ODataBatchOperationResponseMessage"/> for writing an operation of a batch response - implementation of the actual functionality.
+        /// </summary>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        /// <returns>The message that can be used to write the response operation.</returns>
+        protected override ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation(string contentId)
+        {
+            this.WritePendingMessageData(true);
+
+            // In responses we don't need to use our batch URL resolver, since there are no cross referencing URLs
+            // so use the URL resolver from the batch message instead.
+            this.CurrentOperationResponseMessage = BuildOperationResponseMessage(
+                this.rawOutputContext.OutputStream);
+
+            this.SetState(BatchWriterState.OperationCreated);
+
+            Debug.Assert(this.CurrentOperationContentId == null, "The Content-ID header is only supported in request messages.");
+
+            // write the operation's start boundary string
+            this.WriteStartBoundaryForOperation();
+
+            // write the headers and request separator line
+            ODataMultipartMixedBatchWriterUtils.WriteResponsePreamble(this.rawOutputContext.TextWriter, changeSetBoundary != null, contentId);
+
+            return this.CurrentOperationResponseMessage;
+        }
+
+        /// <summary>
+        /// Writes all the pending headers and prepares the writer to write a content of the operation.
+        /// </summary>
+        protected override void StartBatchOperationContent()
+        {
+            Debug.Assert(this.CurrentOperationMessage != null, "Expected non-null operation message!");
+            Debug.Assert(this.rawOutputContext.TextWriter != null, "Must have a batch writer!");
+
+            // write the pending headers (if any)
+            this.WritePendingMessageData(false);
+
+            // flush the text writer to make sure all buffers of the text writer
+            // are flushed to the underlying async stream
+            this.rawOutputContext.TextWriter.Flush();
+        }
+
+        /// <summary>
+        /// Additional processing required when setting a new writer state.
+        /// </summary>
+        /// <param name="newState">The writer state to transition into.</param>
+        protected override void SetStateImplementation(BatchWriterState newState)

```

Comment:
>SetStateImplementation [](start = 32, length = 22)

Why do we need a SetStateImplementation?  SetState is called by the derived class in order to tell the base class that it is transitioning to a new state.  SetStateImplementation would just be a way for the base class to call back into the derived class to tell the derived class that it is transitioning to a new state.  Why doesn't the derived class just call it's private logic in SetStateImplementation immediately before or after calling SetState?

Reply:
Currently the derived class invokes SetState() in 9 places, don't want to add SetStateImplementation for each of them. Also this way SetState keeps multiple things needs to be done together and matches the API doc description.

---
In reply to: [143600602](https://github.com/OData/odata.net/pull/960#discussion_r143600602) [](ancestors = 143600602)

Reply:
Your SetStateImplementation only cares about two state transitions; changesetstarted and changesetcompleted. Each of these transitions is the result of only one place where the derived class calls SetState; in WriteStartChangesetImplementation you call SetState(ChangesetStarted); this is the natural place to call this.changeSetBoundary = ODataMultipartMixedBatchWriterUtils.CreateChangeSetBoundary(this.rawOutputContext.WritingResponse) in the derived class.  In WriteEndChangesetImplementation you call SetState(ChangesetCompleted).  This is the natural place to call this.changeSetBoundary = null in the derived class.  This avoids adding an extra protected method that is called for each state transition to tell the derived class that it just called SetState.


---
In reply to: [143605346](https://github.com/OData/odata.net/pull/960#discussion_r143605346) [](ancestors = 143605346,143600602)